CMakeLists.txt

```diff
@@ -421,8 +421,42 @@ endif()
 
 if(WIN32)
   set(_gRPC_ALLTARGETS_LIBRARIES ${_gRPC_ALLTARGETS_LIBRARIES} ws2_32 crypt32)
+  set(_gRPC_STATIC_WIN32 STATIC)
 endif()
 
+if(BUILD_SHARED_LIBS AND WIN32)
+
+# Currently for shared lib on Windows (i.e. a DLL) certain bits of source code
+# are generated from protobuf definitions by upbc. This source code does not include
+# annotations needed to export these functions from gprc.lib so we have to
+# re-include a small subset of these.
+#
+# This is not an ideal situation because these functions will be unavailable
+# to clients of grpc and the libraries that need this (e.g. grpc++) will
+# include redundant duplicate code. Hence, the duplication is only activated
+# for DLL builds - and should be completely removed when source files are
+# generated with the necessary __declspec annotations.
+set(gRPC_UPB_GEN_DUPL_SRC
```

Comment:
ok, patching the build with gRPC_UPB_GEN_DUPL_SRC and gRPC_ADDITIONAL_WIN_DLL_SRC seems very dirty so we should really have a clear plan how to handle these in the future cleanly, otherwise we're introducing some serious and hard-to-understand technical debt.

If there's a clear (and commited) plan how to get rid of these hacks in the future I might be ok with having there temporarily, but otherwise this is difficult to accept.

Reply:
> ok, patching the build with gRPC_UPB_GEN_DUPL_SRC and gRPC_ADDITIONAL_WIN_DLL_SRC seems very dirty so we should really have a clear plan how to handle these in the future cleanly, otherwise we're introducing some serious and hard-to-understand technical debt.
> 
> If there's a clear (and commited) plan how to get rid of these hacks in the future I might be ok with having there temporarily, but otherwise this is difficult to accept.

Agreed - and it gets worse when we get the tests to build as well.

Question: is the upb-generated and upbdefs-generated code part of the public API?

As things stand, there's no way of annotating these generated files with dllimport/dllexport tags as the capability to do that in upb doesn't seem to exist (though it does in protobuf), so they are not currently exposed - hence the need to duplicate. 

I think the right solution is to factor out all this generated code from being included directly into grpc and into a separate always-static library, say upb_generated. The existing grpc library then gets renamed to something like grpc_base and the grpc definition simply includes grpc_base and the definition for grpc simply includes upb_generated and grpc_base - yielding a library with the exact same contents as the existing grpc, but allowing the duplicated source to be avoided by having those libs that need it because they depend on a grpc in which the symbols are not visible, to depend on grpc_base instead and mix in the symbols at the 'leaf' level.

I think that would work provided the upb-generated symbols are not part of the public API.

Reply:
Future plan is to have a DLL suport from upb and I'm following up on this but it will take sometime as upb needs to create their cmakefile first. Once upb has a proper cmake with DLL support, we'll be able to remove this hack. Until we have it, I'm fine with this workaround.

Reply:
> Future plan is to have a DLL suport from upb and I'm following up on this but it will take sometime as upb needs to create their cmakefile first. Once upb has a proper cmake with DLL support, we'll be able to remove this hack. Until we have it, I'm fine with this workaround.

@veblush @jtattermusch @h-vetinari Would it be possible to arrange a zoom call to discuss options on this? I can see that upb is in flux (I see it just got moved to be a direct part of protobuf from being a repo in its own right), but I don't believe that the proposed changes in this PR add significant extra work with respect to adapting to that change. I'm also not seeing a future in which it won't be necessary at some level to provide additional metadata above and beyond what can be gleaned from the Bazel BUILD file, though I totally agree that such knowledge does not belong in the CMakeLists template file.

Reply:
Currently I'm chasing upb to have DLL support properly and desirable state is to have proper DLL support so that gRPC can get rid of this hack. Until then, I think we can live with this.